activerecord/lib/active_record/validations/uniqueness.rb

```diff
@@ -64,6 +66,34 @@ def find_finder_class_for(record)
         class_hierarchy.detect { |klass| !klass.abstract_class? }
       end
 
+      def validation_needed?(klass, record, attribute)
+        return true if options[:conditions] || options.key?(:case_sensitive)
+
+        scope = Array(options[:scope]).map(&:to_s)
+        attributes = scope + [attribute.to_s]
+
+        return true if attributes.any? { |attribute| record.class._reflect_on_association(attribute) }
+        return true if attributes.any? { |attribute| record.attribute_changed?(attribute) ||
+                                                     record.read_attribute(attribute).nil? }
+
+        !covered_by_unique_index?(klass, attribute, scope)
+      end
+
+      def covered_by_unique_index?(klass, attribute, scope)
+        return @covered_by_unique_index[attribute] if defined?(@covered_by_unique_index)
+
+        @covered_by_unique_index = self.attributes.index_with do |attribute|
```

Comment:
```suggestion
        @covered_by_unique_index ||= self.attributes.index_with do |attribute|
```

Reply:
Also it's probably doesn't matter that much, but instead of `index_with`, you could do:

```ruby
@covered ||= attributes.select { |attr| ... }.to_set
@covered.include?(attribute)
```